barretenberg/cpp/src/barretenberg/vm2/simulation/events/bitwise_event.hpp

```diff
@@ -1,17 +1,40 @@
 #pragma once
 
+#include <cstdint>
+
 #include "barretenberg/vm2/common/constants.hpp"
 #include "barretenberg/vm2/common/memory_types.hpp"
 
-#include <cstdint>
-
 namespace bb::avm2::simulation {
 
+enum class BitwiseTagError {
+    TAG_MISMATCH,
+    INVALID_FF_TAG,
+};
+inline std::string to_string(BitwiseTagError e)
+{
+    switch (e) {
+    case BitwiseTagError::TAG_MISMATCH:
+        return "TAG_MISMATCH";
+    case BitwiseTagError::INVALID_FF_TAG:
+        return "INVALID_FF_TAG";
+    }
+    // We should be catching all the cases above.
+    __builtin_unreachable();
+}
+
+class BitwiseException : public std::runtime_error {
+  public:
+    BitwiseException()
+        : std::runtime_error("Bitwise Exception")
+    {}
+};
+
 struct BitwiseEvent {
     BitwiseOperation operation;
     MemoryValue a;
     MemoryValue b;
-    MemoryValue res;
+    uint128_t res = 0;
```

Comment:
more efficient than a memory value since: it inherits the tag from a and it cannot be an FF